File: crabzilla-example1/crabzilla-example1-core/src/main/java/crabzilla/example1/customer/CustomerData.java
package crabzilla.example1.customer;

import crabzilla.model.DomainEvent;
import crabzilla.model.EntityCommand;
import crabzilla.model.EntityId;
import lombok.Value;

import java.time.Instant;
import java.util.UUID;

public class CustomerData {

  @Value
  public static class CustomerId implements EntityId {
    String stringValue;
  }

  // commands

  @Value
  public static class CreateCustomer implements EntityCommand {
    UUID commandId;
    CustomerId targetId;
    String name;
  }

  @Value
  public static class ActivateCustomer implements EntityCommand {
    UUID commandId;
    CustomerId targetId;
    String reason;
  }

  @Value
  public static class CreateActivateCustomer implements EntityCommand {
    UUID commandId;
    CustomerId targetId;
    String name;
    String reason;
  }


  @Value
  public static class DeactivateCustomer implements EntityCommand {
    UUID commandId;
    CustomerId targetId;
    String reason;
  }

  // events

  @Value
  public static class CustomerActivated implements DomainEvent {
    String reason;
    Instant when;
  }

  @Value
  public static class CustomerCreated implements DomainEvent {
    CustomerId id;
    String name;
  }

  @Value
  public static class CustomerDeactivated implements DomainEvent {
    String reason;
    Instant when;
  }

}

File: crabzilla-kotlin-example1/crabzilla-kotlin-example1-core/src/main/java/io/github/crabzilla/example1/customer/CustomerData.kt
package io.github.crabzilla.example1.customer

import com.fasterxml.jackson.annotation.JsonProperty
import io.github.crabzilla.core.DomainEvent
import io.github.crabzilla.core.entity.EntityId
import io.github.crabzilla.example1.KotlinEntityCommand
import java.time.Instant
import java.util.*

data class CustomerId(val id: String) : EntityId {
  override fun stringValue(): String {
    return id
  }
}

// tag::events[]

data class CustomerCreated(val id: CustomerId, val name: String) : DomainEvent

data class CustomerActivated(val reason: String, val _when: Instant) : DomainEvent

data class CustomerDeactivated(val reason: String, val _when: Instant) : DomainEvent

// end::events[]

// tag::commands[]


data class CreateCustomer(@JsonProperty("commandId") override val _commandId: UUID,
                          @JsonProperty("targetId") override val _targetId: CustomerId,
                          val name: String) : KotlinEntityCommand

data class ActivateCustomer(@JsonProperty("commandId") override val _commandId: UUID,
                            @JsonProperty("targetId") override val _targetId: CustomerId,
                            val reason: String) : KotlinEntityCommand

data class DeactivateCustomer(@JsonProperty("commandId") override val _commandId: UUID,
                              @JsonProperty("targetId") override val _targetId: CustomerId,
                              val reason: String) : KotlinEntityCommand

data class CreateActivateCustomer(@JsonProperty("commandId") override val _commandId: UUID,
                                  @JsonProperty("targetId") override val _targetId: CustomerId,
                                  val name: String, val reason: String) : KotlinEntityCommand

// end::commands[]

data class UnknownCommand(@JsonProperty("commandId") override val _commandId: UUID,
                          @JsonProperty("targetId") override val _targetId: CustomerId)
    : KotlinEntityCommand

